README.md
# django-unslashed

[![CircleCI](https://circleci.com/gh/frnhr/django-unslashed/tree/master.svg?style=shield)](https://circleci.com/gh/frnhr/django-unslashed/tree/master)
[![codecov](https://codecov.io/gh/frnhr/django-unslashed/branch/master/graph/badge.svg)](https://codecov.io/gh/frnhr/django-unslashed)

This middleware provides the inverse of the Django CommonMiddleware `APPEND_SLASH` feature. It can automatically remove trailing URL slashes and 301 redirect to the non-slash-terminated URL. This behavior is performed if the initial URL ends in a slash and is invalid, removing the trailing slash produces a valid URL, and `REMOVE_SLASH` is set to True. Otherwise there is no effect.

For example, foo.com/bar/ will be redirected to foo.com/bar if you don't have a valid URL pattern for foo.com/bar/ but do have a valid pattern for foo.com/bar and `REMOVE_SLASH=True`.


## Fork info

This is a fork from https://github.com/harshvb7/django-unslashed, which is in turn a fork of https://github.com/dghubble/django-unslashed.

Changes to this fork:
    * updated for Django 1.10, compatible down to 1.8, and Python 2.7, 3.3-3.5.
    * PyPI package name with prefix `tocka-`
    * Added CircleCI and Codecov, removed TravisCI    


## Install

To install `django-unslashed`,

```
pip install tocka-django-unslashed
```

If you're using a `requirements.txt` file, add `django-unslashed>=0.3.0` to it.


## Usage

Modify your Django `settings.py` file to add `unslashed.middleware.RemoveSlashMiddleware`
to your `MIDDLEWARE_CLASSES` just before or after `django.middleware.common.CommonMiddleware`.

```python
MIDDLEWARE_CLASSES = (
    # ...
    'unslashed.middleware.RemoveSlashMiddleware',
    'django.middleware.common.CommonMiddleware',
    # ...
)
```

Set `REMOVE_SLASH` to True and `APPEND_SLASH` to False,

```python
APPEND_SLASH = False
REMOVE_SLASH = True
```

If `REMOVE_SLASH` is False or unset, the RemoveSlashMiddleware has no effect.


Set `UNSLASHED_USE_302_REDIRECT` to True if you want to use HttpRedirect instead of HttpPermanentRedirect,

```python
UNSLASHED_USE_302_REDIRECT = True
```

## Rationale

Web applications [should](http://googlewebmastercentral.blogspot.com/2010/04/to-slash-or-not-to-slash.html) have a URL structure which either:

1. Uses trailing slashes and redirects to append slashes if invalid non-slashed-terminated URLs are accessed.
2. Uses no trailing slash URLs and redirects to unslahed URLs if invalid slash terminated URLs are accessed. The prior is the Django default, while the later is possible by adding this middleware to your project.


## Notes

Based closely on Django's APPEND_SLASH CommonMiddleware [implementation](https://github.com/django/django/blob/master/django/middleware/common.py).


## Testing

### Tox

This will run tests on multiple versions of Python and Django, as defined in `tox.ini`.

```bash
$ git clone https://github.com/frnhr/django-unslashed.git
$ cd django-unslashed
$ pip install -r requirements-dev.txt
$ tox
...
...
...
________________________________________________________________________________________________________ summary _________________________________________________________________________________________________________
  django_master-py35: commands succeeded
  django_master-py34: commands succeeded
  django_master-py27: commands succeeded
  django110-py35: commands succeeded
  django110-py34: commands succeeded
  django110-py27: commands succeeded
  django19-py35: commands succeeded
  django19-py34: commands succeeded
  django19-py27: commands succeeded
  django18-py35: commands succeeded
  django18-py34: commands succeeded
  django18-py33: commands succeeded
  django18-py27: commands succeeded
  lint: commands succeeded
  congratulations :)
```


## License

[MIT License](LICENSE)


CHANGES.rst
Changelog
=========

Notable changes between releases.

0.4.0 (2016-12-12)
------------------

* Update for Django 1.10
* Drop support for Python 2.6 and Django 1.6 and 1.7

0.3 (2014-12-18)
----------------

* Fix bug in which encoded characters were re-encoded upon redirection
* Add explicit test coverage for Django 1.6 and 1.7 and Python 2.6 to 3.4
* Require Django >= 1.5

0.2 (2013-12-18)
----------------

* Initial release

unslashed/tests.py
from django.http.response import (
    HttpResponse,
    HttpResponsePermanentRedirect,
    HttpResponseRedirect,
    HttpResponseRedirectBase,
)
from django.test import TestCase, Client
from django.utils.http import urlquote


class RemoveSlashMiddlewareTest(TestCase):
    def setUp(self):
        self.client = Client()

    def test_permanent_redirect_to_non_slashed(self):
        response = self.client.get('/testapps/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

        response = self.client.get('/testapps/1/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))

    def test_temp_redirect_to_non_slashed(self):
        with self.settings(UNSLASHED_USE_302_REDIRECT=True):
            response = self.client.get('/testapps/', follow=False)
            self.assertIsInstance(response, HttpResponseRedirect)
            self.assertFalse(response['Location'].endswith('/'))

            response = self.client.get('/testapps/1/', follow=False)
            self.assertIsInstance(response, HttpResponseRedirect)
            self.assertFalse(response['Location'].endswith('/'))

    def test_whitelisted_slashed(self):
        with self.settings(UNSLASHED_WHITELIST_STARTSWITH=['/testapps/']):
            response = self.client.get('/testapps/', follow=False)
            self.assertNotIsInstance(response, HttpResponseRedirectBase)
            self.assertEqual(response.status_code, 404)

            response = self.client.get('/testapps/1/', follow=False)
            self.assertNotIsInstance(response, HttpResponseRedirectBase)
            self.assertEqual(response.status_code, 404)

    def test_permanent_redirect_to_unslashed_when_url_has_plus_signs(self):
        response = self.client.get('/testapps/url+with+plus/', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus$')

        response = self.client.get('/testapps/url+with+plus/?param1=1&param2=%2Bfoo', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/url\+with\+plus\?param1=1&param2=%2Bfoo$')

    def test_permanent_redirect_to_unslashed_when_url_has_urlencoded_chars(self):
        # urlquote is used because django.test.Client decodes the URL
        # which goes against the test
        slashed_url = urlquote('/testapps/quoted/foo%2Bbar%23baz%20/')
        response = self.client.get(slashed_url, follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertFalse(response['Location'].endswith('/'))
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20$')

        response = self.client.get(slashed_url + '?param1=1&param2=%2Btest', follow=False)
        self.assertIsInstance(response, HttpResponsePermanentRedirect)
        self.assertRegexpMatches(response['Location'], r'/quoted/foo%2Bbar%23baz%20\?param1=1&param2=%2Btest$')

    def test_no_redirect_when_slash_url_is_valid(self):
        response = self.client.get('/testapps/1/urlendsinslash/', follow=False)
        self.assertIsInstance(response, HttpResponse)
        self.assertEqual(response.status_code, 200)

    def test_no_redirect_when_slashed_and_unslashed_invalid(self):
        response = self.client.get('/testapps/invalid/', follow=False)
        self.assertNotIsInstance(response, HttpResponseRedirectBase)
        self.assertEqual(response.status_code, 404)

        response = self.client.get('/testapps/invalid', follow=False)
        self.assertNotIsInstance(response, HttpResponseRedirectBase)
        self.assertEqual(response.status_code, 404)

    def test_warns_about_redirect_and_post(self):
        with self.settings(DEBUG=True):
            with self.assertRaises(RuntimeError):
                self.client.post('/testapps/', follow=False)

    def test_disabled_middleware(self):
        with self.settings(REMOVE_SLASH=False):
            response = self.client.get('/testapps/', follow=False)
            self.assertNotIsInstance(response, HttpResponseRedirectBase)
            self.assertEqual(response.status_code, 404)



unslashed/middleware.py
import re
from django.conf import settings
from django.core import urlresolvers
from django.core.exceptions import MiddlewareNotUsed
from django.http import HttpResponsePermanentRedirect, HttpResponseRedirect
from django.utils.encoding import iri_to_uri
try:
    from django.utils.deprecation import MiddlewareMixin
except ImportError:  # pragma: no cover
    MiddlewareMixin = object  # Django<1.10


trailing_slash_regexp = re.compile(r'(\/(?=\?))|(\/$)')


class RemoveSlashMiddleware(MiddlewareMixin):
    """
    This middleware provides the inverse of the APPEND_SLASH option built into
    django.middleware.common.CommonMiddleware. It should be placed just before
    or just after CommonMiddleware.

    If REMOVE_SLASH is True, the initial URL ends with a slash, and it is not
    found in the URLconf, then a new URL is formed by removing the slash at the
    end. If this new URL is found in the URLconf, then Django redirects the
    request to this new URL. Otherwise, the initial URL is processed as usual.

    For example, foo.com/bar/ will be redirected to foo.com/bar if you don't
    have a valid URL pattern for foo.com/bar/ but do have a valid pattern for
    foo.com/bar.

    Using this middleware with REMOVE_SLASH set to False or without
    REMOVE_SLASH set means it will do nothing.

    Originally, based closely on Django's APPEND_SLASH CommonMiddleware
    implementation at
    https://github.com/django/django/blob/master/django/middleware/common.py.
    It has been reworked to use regular expressions instead of deconstructing/
    reconstructing the URL, which was problematically re-encoding some of the
    characters.
    """

    def __init__(self, *args, **kwargs):
        if not getattr(settings, 'REMOVE_SLASH', False):
            raise MiddlewareNotUsed()
        self.response_class = HttpResponsePermanentRedirect
        if getattr(settings, 'UNSLASHED_USE_302_REDIRECT', None):
            self.response_class = HttpResponseRedirect
        super(RemoveSlashMiddleware, self).__init__(*args, **kwargs)

    def should_redirect_without_slash(self, request):
        """
        Return True if appending a slash to
        the request path turns an invalid path into a valid one.
        """
        if trailing_slash_regexp.search(request.get_full_path()):
            urlconf = getattr(request, 'urlconf', None)

            return (
                not urlresolvers.is_valid_path(request.path_info, urlconf) and
                urlresolvers.is_valid_path(request.path_info[:-1], urlconf))
        return False

    def get_full_path_without_slash(self, request):
        """
        Return the full path of the request without a trailing slash appended.

        Raise a RuntimeError if settings.DEBUG is True and request.method is
        POST, PUT, or PATCH.
        """
        path = request.path
        if path.endswith('/'):
            path = path[:-1]
        new_path = '{}{}'.format(
            path,
            ('?' + iri_to_uri(request.META.get('QUERY_STRING', '')))
            if request.META.get('QUERY_STRING', '')
            else ''
        )
        if settings.DEBUG and request.method in ('POST', 'PUT', 'PATCH'):
            raise RuntimeError(
                "You called this URL via %(method)s, but the URL ends "
                "in a slash and you have REMOVE_SLASH set. Django-unslashed "
                "can't redirect to the slash URL while maintaining %(method)s "
                "data. Change your form to point to %(url)s (note there is no"
                "trailing slash), or set REMOVE_SLASH=False in your Django "
                " settings." % {
                    'method': request.method,
                    'url': request.get_host() + new_path,
                }
            )
        return new_path

    def process_response(self, request, response):
        """
        Redirects to the current URL without the trailing slash if
        settings.REMOVE_SLASH is True and our current response's
        status_code would be a 404
        """
        # If the given URL is "Not Found", then check if we should redirect to
        # a path without a slash appended.
        if response.status_code == 404:
            path = request.path
            whitelists = getattr(settings,
                                 'UNSLASHED_WHITELIST_STARTSWITH',
                                 ['/admin'])
            if any(path.startswith(x) for x in whitelists):
                return response
            if self.should_redirect_without_slash(request):
                return self.response_class(
                    self.get_full_path_without_slash(request))
        return response
